invokeai/app/services/graph.py

```diff
@@ -377,53 +336,108 @@ def delete_edge(self, edge: Edge) -> None:
         except KeyError:
             pass
 
-    def is_valid(self) -> bool:
-        """Validates the graph."""
+    def validate_self(self) -> None:
+        """
+        Validates the graph.
+
+        Raises an exception if the graph is invalid:
+        - `DuplicateNodeIdError`
+        - `NodeIdMismatchError`
+        - `InvalidSubGraphError`
+        - `NodeNotFoundError`
+        - `NodeFieldNotFoundError`
+        - `CyclicalGraphError`
+        - `InvalidEdgeError`
+        """
+
+        # Validate that all node ids are unique
+        node_ids = [n.id for n in self.nodes.values()]
```

Comment:
if len(node_ids) > 1, should all nodes be on an edge?

Reply:
No, you can have a graph with N nodes and zero edges

Reply:
It's allowed - ex a single resize image node

Reply:
note that all of this validation logic already existed, it's just been pulled into a single function now and called when the session/batch is created. every session within the batch is also validated. should be fairly difficult to get invalid graphs in

Reply:
Weird I looked at this today and didn't see any comments but I can see them on my phone